### PR TITLE
perf: pre-compile embedded Lua source into bytecode

### PR DESCRIPTION
**Pull request by gpanders**
Currently, the Lua modules that make up vim.lua are embedded as raw source
files into the `nvim` binary. These sources are loaded by the Lua runtime on
startuptime. We can pre-compile these sources into Lua bytecode before
embedding them into the binary, which minimizes the size of the binary and
improves startuptime.

The benefit of this is greater for larger modules (such as the proposed `filetype.lua` #16600).

### CODE REVIEW

File: src/nvim/CMakeLists.txt [hunk 1]
```diff
@@ -326,7 +326,9 @@ add_custom_command(
 
 add_custom_command(
   OUTPUT ${VIM_MODULE_FILE}
-  COMMAND ${LUA_PRG} ${CHAR_BLOB_GENERATOR} ${VIM_MODULE_FILE}
+  COMMAND ${CMAKE_COMMAND} -E env
+      "LUAC_PRG=${LUAC_PRG}"
+      ${LUA_PRG} ${CHAR_BLOB_GENERATOR} ${VIM_MODULE_FILE}
```
Comment (gpanders) [inline src/nvim/CMakeLists.txt:331]: Windows runners don't seem to like this and complain about `cmake -E env: No command given`. Anyone with more Cmake knowledge know how to address this? (poke @jamessan).

Comment (jamessan) [reply src/nvim/CMakeLists.txt:331]: I think we need someone with a Windows dev environment setup to test this and see what's going wrong.  @erw7?

File: src/nvim/generators/gen_char_blob.lua [hunk 2]
```diff
@@ -28,6 +28,8 @@ for argi = 2, #arg, 2 do
 
   target:write(('static const uint8_t %s[] = {\n'):format(varname))
 
+  local compiled = assert(io.popen(os.getenv("LUAC_PRG"):format(source_file)):read("*a"))
```
Comment (bfredl) [inline src/nvim/generators/gen_char_blob.lua:31]: `gen_char_blob` could be used for other things than lua source potentially. perhaps add an `arg` to toggle compilation?

File: src/nvim/CMakeLists.txt (hunk 1 above)
Comment (erw7) [reply src/nvim/CMakeLists.txt]: I think we need to do some escaping. I have tried several things, but they do not work. On Windows, how about using `loadstring` and `string.dump` as follows? This method is about `3k` larger (`vim.lua`: 20k, pre-compile: 8k, this method: 11k), but the load time doesn't seem to change much.

```lua
local compiled = string.dump(assert(loadstring(sourc:read('*a'), source_name)))
```

Comment (gpanders) [reply src/nvim/CMakeLists.txt]: It looks like https://github.com/neovim/neovim/pull/16631/commits/a1492c830c69950a4255df9614dac81c7540b14a solved the first problem, but now there are other errors :)

**Comment by gpanders** [comment]
Quite a few Windows failures and they all seem somewhat random? I'm not sure why pre-compiling the byte code would cause these failures at least.

Any objection to only pre-compiling for Unix-es? We can add a guard in `gen_char_blob` (or CMakeLists.txt) based on OS.

**Comment by gpanders** [comment]
Conditionally compiling only on non-Windows platforms seems to work

**Comment by kylo252** [comment]
> Quite a few Windows failures and they all seem somewhat random? I'm not sure why pre-compiling the byte code would cause these failures at least.

It's not related to `local luac = os.getenv("LUAC_PRG")` ? I'm not sure if this is reliable enough on Windows. Let me know if you want to me try something.

**Comment by gpanders** [comment]
> > Quite a few Windows failures and they all seem somewhat random? I'm not sure why pre-compiling the byte code would cause these failures at least.
> 
> It's not related to `local luac = os.getenv("LUAC_PRG")` ? I'm not sure if this is reliable enough on Windows. Let me know if you want to me try something.

That was only added in response to the Windows failures.

If you want, you can check out https://github.com/neovim/neovim/pull/16631/commits/e9a0bd532531bf630ef9b9ca25a2984680141900 and run the [tests that failed](https://github.com/neovim/neovim/runs/4523094350?check_suite_focus=true) locally and see if you can figure out what's going on.

I don't think this should be a blocker for this PR though. Providing this feature for Unix-es first is already a win. We can come back and add Windows support in a follow up PR.

**Comment by kylo252** [comment]
> > > Quite a few Windows failures and they all seem somewhat random? I'm not sure why pre-compiling the byte code would cause these failures at least.
> > 
> > 
> > It's not related to `local luac = os.getenv("LUAC_PRG")` ? I'm not sure if this is reliable enough on Windows. Let me know if you want to me try something.
> 
> That was only added in response to the Windows failures.
> 
> If you want, you can check out [e9a0bd5](https://github.com/neovim/neovim/commit/e9a0bd532531bf630ef9b9ca25a2984680141900) and run the [tests that failed](https://github.com/neovim/neovim/runs/4523094350?check_suite_focus=true) locally and see if you can figure out what's going on.

So just from a quick overview, I think this might be the issue:
```console
# windows
# "cmake: Using Lua compiler: <neovim-repo>/.deps/usr/bin/luajit.exe -b -s %s - "
> .\.deps\usr\bin\luajit.exe -b --help
.deps\usr\bin\luajit.exe: unknown luaJIT command or jit.* modules not installed
```
vs
```console
# linux
❯ luajit -b --help
Save LuaJIT bytecode: luajit -b[options] input output
  -l        Only list bytecode.
  -s        Strip debug info (default).
  -g        Keep debug info.
  -n name   Set module name (default: auto-detect from input name).
  -t type   Set output file type (default: auto-detect from output name).
  -a arch   Override architecture for object files (default: native).
  -o os     Override OS for object files (default: native).
  -e chunk  Use chunk string as input.
  --        Stop handling options.
  -         Use stdin as input and/or stdout as output.

File types: c h obj o raw (default)
```

**Comment by gpanders** [comment]
I guess Windows doesn't use dashes for options, they use slashes.

Does `luajit /b` work?

**Comment by kylo252** [comment]
> I guess Windows doesn't use dashes for options

Not necessarily, as that is more specific to batch.

```console
❯ .\.deps\usr\bin\luajit.exe -v
LuaJIT 2.1.0-beta3 -- Copyright (C) 2005-2021 Mike Pall. https://luajit.org/

> .\.deps\usr\bin\luajit.exe --help
usage: 
.\.deps\usr\bin\luajit.exe [options]... [script [args]...].
Available options are:
  -e chunk  Execute string 'chunk'.
  -l name   Require library 'name'.
  -b ...    Save or list bytecode.
  -j cmd    Perform LuaJIT control command.
  -O[opt]   Control LuaJIT optimizations.
  -i        Enter interactive mode after executing 'script'.
  -v        Show version information.
  -E        Ignore environment variables.
  --        Stop handling options.
  -         Execute stdin and stop handling options.
```

**Comment by kylo252** [comment]
I think you can see the error here: https://github.com/neovim/neovim/runs/4523094350?check_suite_focus=true#step:4:510

**Comment by gpanders** [comment]
Let's table the problems with Windows for now and take this one step at a time. We can address that in a follow-up.

**Comment by erw7** [comment]
If you install the `jit` module in the next patch, it will compile on Windows. But, there is no point in spending a lot of time on this PR. It is true that precompiled binaries load about 300 times faster, but the time taken to load is less than one millisecond, so a 300x speedup is not expected to have a noticeable effect on boot time.

```diff
--git a/src/nvim/generators/gen_char_blob.lua b/src/nvim/generators/gen_char_blob.lua
index fa885a788..ed9310c61 100644
--- a/src/nvim/generators/gen_char_blob.lua
+++ b/src/nvim/generators/gen_char_blob.lua
@@ -12,7 +12,7 @@ end
 local options = {}
 
 while true do
-  local opt = string.match(arg[1], "-(%w)")
+  local opt = string.match(arg[1], "^-(%w)")
   if not opt then
     break
   end
diff --git a/third-party/cmake/BuildLuajit.cmake b/third-party/cmake/BuildLuajit.cmake
index c4e5112dc..071a3f5d9 100644
--- a/third-party/cmake/BuildLuajit.cmake
+++ b/third-party/cmake/BuildLuajit.cmake
@@ -124,7 +124,8 @@ elseif(MINGW)
 	    COMMAND ${CMAKE_COMMAND} -E copy ${DEPS_BUILD_DIR}/src/luajit/src/libluajit.a ${DEPS_INSTALL_DIR}/lib
 	    COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_INSTALL_DIR}/include/luajit-2.1
 	    COMMAND ${CMAKE_COMMAND} -DFROM_GLOB=${DEPS_BUILD_DIR}/src/luajit/src/*.h -DTO=${DEPS_INSTALL_DIR}/include/luajit-2.1 -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/CopyFilesGlob.cmake
-          )
+	    COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_INSTALL_DIR}/bin/lua/jit
+	    COMMAND ${CMAKE_COMMAND} -E copy_directory ${DEPS_BUILD_DIR}/src/luajit/src/jit ${DEPS_INSTALL_DIR}/bin/lua/jit)
 elseif(MSVC)
 
   BuildLuaJit(
@@ -138,7 +139,9 @@ elseif(MSVC)
       # Luv searches for luajit.lib
       COMMAND ${CMAKE_COMMAND} -E copy ${DEPS_BUILD_DIR}/src/luajit/src/lua51.lib ${DEPS_INSTALL_DIR}/lib/luajit.lib
       COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_INSTALL_DIR}/include/luajit-2.1
-      COMMAND ${CMAKE_COMMAND} -DFROM_GLOB=${DEPS_BUILD_DIR}/src/luajit/src/*.h -DTO=${DEPS_INSTALL_DIR}/include/luajit-2.1 -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/CopyFilesGlob.cmake)
+      COMMAND ${CMAKE_COMMAND} -DFROM_GLOB=${DEPS_BUILD_DIR}/src/luajit/src/*.h -DTO=${DEPS_INSTALL_DIR}/include/luajit-2.1 -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/CopyFilesGlob.cmake
+	    COMMAND ${CMAKE_COMMAND} -E make_directory ${DEPS_INSTALL_DIR}/bin/lua/jit
+	    COMMAND ${CMAKE_COMMAND} -E copy_directory ${DEPS_BUILD_DIR}/src/luajit/src/jit ${DEPS_INSTALL_DIR}/bin/lua/jit)
 
 else()
   message(FATAL_ERROR "Trying to build luajit in an unsupported system ${CMAKE_SYSTEM_NAME}/${CMAKE_C_COMPILER_ID}")
```

**Comment by kylo252** [comment]
> If you install the jit module in the next patch, it will compile on Windows.

I still think that this fix should go in a separate PR. I remember coming across this issue before, so it would be better to be able to reference it in the future imo.

**Comment by erw7** [comment]
> I still think that this fix should go in a separate PR. I remember coming across this issue before, so it would be better to be able to reference it in the future imo.

Is it the same problem as #15543 you are talking about? If so, that's another matter. #15543 resolution requires installing the `jit` module in `${CMAKE_INSTALL_PREFIX}`, not `${DEPS_INSTALL_DIR}/bin/lua/jit`. It is only for this PR that we need to install it in `${DEPS_INSTALL_DIR}/bin/lua/jit`.

**Comment by erw7** [comment]
On Windows, `'rb'` is required, as shown below. However, for some reason, `LF` is converted to `CRLF`, so the binary will say "cannot load malformed bytecode" and terminate abnormally. If there is no `'rb'`, the input will be terminated midway at `26 (0x1A)`. I can't help you. If you really need this for Windows, please try your best.

```diff
diff --git a/src/nvim/generators/gen_char_blob.lua b/src/nvim/generators/gen_char_blob.lua
index fa885a788..650523ad7 100644
@@ -44,7 +44,7 @@ for argi = 2, #arg, 2 do

   local output
   if options.c then
-    output = assert(io.popen(os.getenv("LUAC_PRG"):format(source_file)):read("*a"))
+    output = assert(io.popen(os.getenv("LUAC_PRG"):format(source_file), 'rb'):read("*a"))
   else
     local f = io.open(source_file)
     output = f:read("*a")
```

<details><summary>vim_module.generated.h with 'rb'</summary>

```C
#include <stdint.h>

static const uint8_t vim_module[] = {
   27,  76,  74,   2,  13,  10, 198,   4,   0,   1,  16,   1,  25,   1, 102,
  ...
```
</details>

<details><summary>vim_module.generated.h without 'rb'</summary>

```C
#include <stdint.h>

static const uint8_t vim_module[] = {
   27,  76,  74,   2,  10, 198,   4,   0,   1,  16,   1,  25,   1, 102,  18,
    3,   0,   0,  57,   1,   0,   0,  39,   4,   1,   0,  39,   5,   2,   0,
   ... 
    2,   1,   0,  39,   4,   7,   0,  66,   2,   2,   2,  61,   2,   6,   0,
   57,   2,   6,   0,  76,   2,   2,   0,  88,   2,  0};
```
</details>

<details><summary>vim_module.generated.h generate in linux</summary>

```C
#include <stdint.h>

static const uint8_t vim_module[] = {
   27,  76,  74,   2,  10, 198,   4,   0,   1,  16,   1,  25,   1, 102,  18,
    3,   0,   0,  57,   1,   0,   0,  39,   4,   1,   0,  39,   5,   2,   0,
  ...
    2,   1,   0,  39,   4,   7,   0,  66,   2,   2,   2,  61,   2,   6,   0,
   57,   2,   6,   0,  76,   2,   2,   0,  88,   2,  26, 128,   7,   1,   8,
```
</details>

**Review by jamessan — APPROVED** [review]

**Review by bfredl — APPROVED** [review]

**Comment by erw7** [comment]
> However, for some reason, LF is converted to CRLF, so the binary will say "cannot load malformed bytecode" and terminate abnormally.

It was a bug in `jit.bcsave`. The fix for LuaJIT/LuaJIT#782. should allow pre-compilation on Windows. Those who need it can create their own follow-up PR.

**Comment by clason** [comment]
@erw7 that fix is now merged in LuaJIT; should we bump our bundled version?

**Comment by shadmansaleh** [comment]
Does this break cross compilation ? Termux build for android seems to be broken after this . I get same error as #16709 on  luajit(master) +  neovim(master) .

**Comment by gpanders** [comment]
I’m not sure. It’s just compiling to byte code, not machine code, so I’d be surprised if that was architecture dependent.

Ideally we’d like to identify all the corner cases where this isn’t working and address the root cause. If that’s something you can help contribute to for the cross compilation case then PRs are extremely appreciated. But in the meantime I will add a feature flag to disable this for build systems that can’t support it for whatever reason.

**Comment by shadmansaleh** [comment]
According to https://www.lua.org/manual/5.1/luac.html

> The binary files created by luac are portable only among architectures with the same word size and byte order

So the byte code might not be always cross compatable ?

**Comment by gpanders** [comment]
@shadmansaleh #16722

**Comment by shadmansaleh** [comment]
Thanks that fixed the issue .

**Pull request merged**